### ISSUE TITLE
[BUG] `cpdef_enums` test broken in Py3.12

### ISSUE THREAD

**Issue by scoder** [cython/cython#5469]
### Describe the bug

```
======================================================================
FAIL: test_pickle (cpdef_enums)
Doctest: cpdef_enums.test_pickle
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/.../python3.12/doctest.py", line 2228, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for cpdef_enums.test_pickle
  File "/.../TEST_TMP/run/c/cpdef_enums/cpdef_enums.cpython-312d-x86_64-linux-gnu.so", line unknown line number, in test_pickle

----------------------------------------------------------------------
File "/.../TEST_TMP/run/c/cpdef_enums/cpdef_enums.cpython-312d-x86_64-linux-gnu.so", line ?, in cpdef_enums.test_pickle
Failed example:
    loads(dumps(PxdEnum.RANK_2)) == PxdEnum.RANK_2
Exception raised:
    Traceback (most recent call last):
      File "/.../python3.12/doctest.py", line 1357, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest cpdef_enums.test_pickle[4]>", line 1, in <module>
        loads(dumps(PxdEnum.RANK_2)) == PxdEnum.RANK_2
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    TypeError: unsupported operand type(s) for |: 'PxdEnum' and 'NoneType'
```

### Python version

3.12.0b1

### Cython version

master

**Comment by da-woods** [comment]
I think this is a Python bug (or at least a Python change that may be intended).

Taking the following Python code (i.e. no Cython compilation at all)

```
import enum
import pickle

class PxdEnum(enum.IntFlag):
    RANK_0 = 11
    RANK_1 = 37
    RANK_2 = 389

print(pickle.loads(pickle.dumps(PxdEnum.RANK_2)))
```

On Python 3.10 I get `PxdEnum.RANK_2`; on Python 3.11 I get `389` (which is different, but probably OK because it compares equal). On Python 3.12 (tip of current main branch) I get

```
Traceback (most recent call last):
  File "<path>/enumpickle.py", line 9, in <module>
    print(pickle.loads(pickle.dumps(PxdEnum.RANK_2)))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unsupported operand type(s) for |: 'PxdEnum' and 'NoneType'
```

**Comment by da-woods** [comment]
Python bug report is https://github.com/python/cpython/issues/105332.

I'll wait to see what the outcome of that is before getting too involved in trying to fix Cython. If it isn't considered a bug we may have to do something though (since I think we'd like pickling/unpickling cpdef enums to work)

**Comment by da-woods** [comment]
Now fixed in the Python 3.12 branch

**Comment by da-woods** [comment]
So it now looks like this is broken in Python 3.11 too... So my bug report possibly wasn't a complete success. I still think this is third party but I'll reopen the issue here

**Comment by da-woods** [comment]
https://github.com/python/cpython/issues/105837

**Comment by da-woods** [comment]
Right - the conclusion was that 3.11.4 is buggy for pickling enums, but it'll be fixed in the next 3.11 release.

**Comment by da-woods** [comment]
Closing this again. We've established the issue is 3rd party, and disabled the tests on the offending versions. So I don't think there's more we can do under this issue
